### DISCUSSION TITLE
"version" field in Type #1 and Type #2 (UKI) entries

### DISCUSSION THREAD

**Discussion by osteffenrh** [uapi-group/specifications#36] [General]
The Boot Loader Spec states ([see here](https://uapi-group.org/specifications/specs/boot_loader_specification/#type-2-efi-unified-kernel-images)):
> The PRETTY_NAME= and VERSION_ID= fields in the embedded os-release file are used the same as title and version in the Type `#1` entries.

For Type `#1` entries the `version` field is used as the kernel version ([see here](https://uapi-group.org/specifications/specs/boot_loader_specification/#type-1-boot-loader-entry-keys)).

BUT: The os-release file and thus the `.osrel` section of the UKI uses the `VERSION_ID` for the OS version (not the kernel version). (see [`man /etc/os-release`](https://man7.org/linux/man-pages/man5/os-release.5.html))

This is also what I see on Fedora 36:
Loader entry:
```
title      Fedora Linux 36 (Server Edition)
version    5.17.5-300.fc36.x86_64
machine-id 9d1099c6ce0146be9bfb6bdde1e42b5e
options    root=UUID=9011d417-655a-4c5a-bb6c-719859a14119 ro rhgb quiet console=tty0 console=ttyS0
linux      /9d1099c6ce0146be9bfb6bdde1e42b5e/5.17.5-300.fc36.x86_64/linux
initrd     /9d1099c6ce0146be9bfb6bdde1e42b5e/5.17.5-300.fc36.x86_64/initrd
```

`.osrel` and `/etc/os-release`:
```
NAME="Fedora Linux"
VERSION="36 (Server Edition)"
ID=fedora
VERSION_ID=36
VERSION_CODENAME=""
PLATFORM_ID="platform:f36"
PRETTY_NAME="Fedora Linux 36 (Server Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:36"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f36/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
```

**Comment by poettering** [comment]
Yeah, this is a bit borked. So I have half a PR ready which adds ".uname" as a new PE section to UKIs. With that we'd have two version codes in a UKI. The distro version in the .osrel field, and the kernel version in .uname.

I figure once that is in place, we should maybe do two things:

1. for type `#2` entries combine the two versions from the two sections, and sort by that, i.e. `$VERSION_ID-$UNAME` or so. 
2. for type `#1` entries we probably should do the same by default now in kernel-install when writing them out.

With that in place I think we should have a good solution: first ordering key is the OS version, second ordering key is the kernel versoin, regardless if type1 or type2.

Does that make sense?

@keszybz?

**Reply by equaeghe** [reply]
It seems to me that the kernel version (from .uname) should be the primary source for ‘version’. It is perfectly possible to use the same kernel for OSs with a different version (e.g., because many other OS components changed). Furthermore, for rolling-release distributions, there is often no sensible OS version. (For example, for Gentoo, the version in os-release is the version number for the package supplying the os-release file, which has no meaning as part of the selection of kernel in the boot menu.)

So I am positive towards changes incorporating .uname, and hope this discussion can be bought back to life, but would like to request to avoid a fixed format in combination with the version from .osrel, i.e., anything that sets in stone the assumption that the .osrel version is meaningful information for the boot menu line for a kernel.
